### ISSUE TITLE
Type narrowing not enforced for object properties with union types

### ISSUE THREAD

**Issue by MathiasWP** [microsoft/TypeScript#61126]
### 🔎 Search Terms

- Type narrowing not enforced for object properties with union types
- TypeScript does not detect unintentional type error"
- Make TypeScript error when type does not match"

### 🕗 Version & Regression Information

I checked all the way back to version 3.3.3, and it didn't behave as expected on any of the versions.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.7.3#code/PQKhCgAIUgVALAlgZ0gEwPYFNUDsMAukB8AThgO6QCGukWp5pUIw4AxhrskSYrgHMAXJADeNALYYArrgIjc0iQCMGAbQC6kAD6RFKhpAC+kALxiokK9Smz5kNQEYN4I+HB9BAOhsy5Z03NHd1AIaEgAdSQAGyxiJFQScgpUWnpGDGZoNk5uIl87BSVVUk0dPWLDcycXcAL-QKCgA

### 💻 Code

```ts
/**
 * This does not throw an error
 */
const thing: { amount: number[] | number } = {
    amount: [1]
}

thing.amount === 1

/**
 * While this throws an error
 */
const amount: number[] | number = [1]

amount === 1
```

### 🙁 Actual behavior

The `thing.amount === 1` does not throw an `This comparison appears to be unintentional because the types 'number[]' and 'number' have no overlap.` error.

### 🙂 Expected behavior

I expect it to throw the `This comparison appears to be unintentional because the types 'number[]' and 'number' have no overlap.` error.
